include/verilated_timing.h

```diff
@@ -0,0 +1,367 @@
+private:
+    // TYPES
+    using DelayedCoro = std::pair<uint64_t, std::coroutine_handle<>>;
+    struct Cmp {
+        bool operator()(const DelayedCoro& lhs, const DelayedCoro& rhs) {
+            return lhs.first > rhs.first;
+        }
+    };
+    using DelayedPriorityQueue = std::priority_queue<DelayedCoro, std::vector<DelayedCoro>, Cmp>;
+
+    // MEMBERS
+    DelayedPriorityQueue m_queue;  // Actual encapsulated queue
+    VerilatedMutex m_mutex;  // Protects m_queue
+
+public:
+    ~VerilatedDelayedQueue() {
+        while (!m_queue.empty()) {
+            m_queue.top().second.destroy();
+            m_queue.pop();
+        }
+    }
+
+    // METHODS
+    // Push a suspended coroutine to be resumed at the given simulation time
+    void push(uint64_t time, std::coroutine_handle<> coro) {
+        const VerilatedLockGuard guard{m_mutex};
+        m_queue.push(std::make_pair(time, coro));
+    }
+
+    // Resume coroutines waiting for the given simulation time
+    void resume(uint64_t time) {
+        VerilatedLockGuard guard{m_mutex};
+        while (!m_queue.empty() && m_queue.top().first <= time) {
+            auto coro = m_queue.top().second;
+            m_queue.pop();
+            guard.unlock();
+            coro();
+            guard.lock();
+        }
+    }
+
+    // Simulation time of the next time slot
+    uint64_t nextTimeSlot() {
+        const VerilatedLockGuard guard{m_mutex};
+        if (!m_queue.empty())
+            return m_queue.top().first;
+        else
+            return VL_TIME_Q();
+    }
+
+    // Is the queue empty?
+    bool empty() {
+        const VerilatedLockGuard guard{m_mutex};
+        return m_queue.empty();
+    }
+
+    // Used by coroutines for co_awaiting a certain simulation time
+    auto operator[](uint64_t time) {
+        const VerilatedLockGuard guard{m_mutex};
+        struct Awaitable {
+            VerilatedDelayedQueue& queue;  // The queue to push the coroutine to
+            uint64_t time;  // Simulation time at which the coroutine should be resumed
+
+            bool await_ready() { return false; }  // Always suspend
+
+            void await_suspend(std::coroutine_handle<> coro) { queue.push(time, coro); }
+
+            void await_resume() {}
+        };
+        return Awaitable{*this, time};
+    }
+};
+
+// Event variable-related type aliases
+using VerilatedEvent = CData;
+using VerilatedEventSet = std::unordered_set<VerilatedEvent*>;
+using VerilatedCoroutineArray = std::vector<std::coroutine_handle<>>;
+
+//=============================================================================
+// VerilatedEventToCoroMap
+// Mapping from events to suspended coroutines. Used by VerilatedEventDispatcher.
+class VerilatedEventToCoroMap final {
+private:
+    // TYPES
+    struct Hash {
+        size_t operator()(const VerilatedEventSet& set) const {
+            size_t result = 0;
+            for (auto event : set) result ^= std::hash<VerilatedEvent*>()(event);
```

Comment:
```suggestion
            for (const auto event : set) result ^= std::hash<VerilatedEvent*>()(event);
```

Many other const's needed in this file, check through please.